cpp/memilio/math/time_series_functor.h

```diff
@@ -0,0 +1,143 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+#ifndef MIO_MATH_TIME_SERIES_FUNCTOR_H
+#define MIO_MATH_TIME_SERIES_FUNCTOR_H
+
+#include "memilio/io/auto_serialize.h"
+#include "memilio/math/interpolation.h"
+#include "memilio/utils/time_series.h"
+
+#include <algorithm>
+#include <cassert>
+#include <vector>
+
+namespace mio
+{
+
+template <class FP>
+class TimeSeriesFunctor
+{
+public:
+    /**
+     * @brief Type of a TimeSeriesFunctor.
+     * The available types are:
+     * - Zero:
+     *   - No data used.
+     *   - Always returns 0.
+     * - LinearInterpolation:
+     *   - Requires at least one time point with exactly one value each. Time must be strictly monotic increasing.
+     *   - Linearly interpolates between data points. Stays constant with first/last value outside of provided data.
+     */
+    enum Type
+    {
+        Zero,
+        LinearInterpolation,
+    };
+
+    /**
+     * @brief Creates a functor using the given data.
+     * Note the data requirements of the given functor type.
+     * @param type The type of the functor.
+     * @param table A list of time points, passed to the TimeSeries constructor.
+     */
+    TimeSeriesFunctor(Type type, const TimeSeries<FP>& data)
+        : m_type(type)
+        , m_data(data)
+    {
+        // data shape checks and preprocessing
+        switch (m_type) {
+        case Type::Zero:
+            // no checks needed
+            break;
+        case Type::LinearInterpolation:
+            // make sure data has the correct shape, i.e. a list of (time, value) pairs
+            assert(m_data.get_num_time_points() > 0 && "Need at least one time point for LinearInterpolation.");
+            assert(m_data.get_num_elements() == 1 && "LinearInterpolation requires exactly one value per time point.");
+            assert(m_data.is_sorted());
+        }
+    }
+
+    /**
+     * @brief Creates a functor using the given table.
+     * Note the data requirements of the given functor type.
+     * @param type The type of the functor.
+     * @param table A list of time points, passed to the TimeSeries constructor.
+     */
+    TimeSeriesFunctor(Type type, std::vector<std::vector<FP>>&& table)
+        : TimeSeriesFunctor(type, TimeSeries<FP>{table})
+    {
+    }
+
+    /**
+     * @brief Creates a Zero functor.
+     */
+    TimeSeriesFunctor()
+        : TimeSeriesFunctor(Type::Zero, TimeSeries<FP>{0})
+    {
+    }
+
+    /**
+     * @brief Function returning a scalar value.
+     * @param time A scalar value.
+     * @return A scalar value computed from data, depending on the functor's type.
+     */
+    FP operator()(FP time) const
+    {
+        FP value = 0.0;
+        switch (m_type) {
+        case Type::Zero:
+            // value is explicitly zero-initialized
```

Comment:
I think this structure is more readable, instead of assigning to a shared value, then you don't need a comment like this.
```
switch(...)
   case Zero:
      return 0.0;
   case Linear:
      return linear_interpolation(...);
   default:
      assert(false && "...");
      return FP();
```

People rightly advise to avoid multiple returns, but this is a good exception.